### DISCUSSION TITLE
Can an application send a message to mastodon user, or in general, to Fediverse user?

### DISCUSSION THREAD

**Discussion by koteisaev** [mastodon/mastodon#26086] [Q&A] [answered]
**Context**
Usually, account creation using a social media app, assumes user would provide an email address, some times as part of requested information. In case of Mastodon, it can be done via separate registration step, with a page/screen asking email address during registration, also.
But if application, created for Fediverse users, can send a Fediverse user a private message with one-time sign-in link, or OTP, it may not require any emails on account at all.
**Question**
Can any arbitrary website send a notification or private message to user, with just information about account obtained via webfinger?

**Comment by ryanmio** [comment, marked as answer]
Sure, you can send messages to Mastodon or other Fediverse users via their API. First, you'd need to OAuth-authenticate your app with their instance. Then, you'd make an API call to send a direct message.

As for using Mastodon instead of email for OTP, it's doable but might be confusing to get OTPs through dms.

And no, you can't just fire off a message with only webfinger info. You'd need OAuth tokens, which means the user has to give your app permission to interact with their account.

**Reply by koteisaev** [reply]
The flow you described, makes Mastodon/ActivityPub a good notifications transport, as you can send notifications only to users who approved that (gave auth token), essentially it means you must provide some value to users that is good enough users would be motivated to authenticate the software to send notifications.

Still, it is not known to me:
* a way to specify a granular scope like "only send direct (mentions-only) messages to user, e. g. without access to other messages user send or receive to/from other accounts"
* parameters for message like "message expire and self-destruct in 24 hours"
Sounds like both require updates to ActivityPub protocol, and deserve own separated issues for discussions, I guess.

**Reply by ryanmio** [reply]
The self-destructing message parameter is a really interesting idea, I could see that being useful for a lot more than OTP codes. I hope that gets some traction.
